### ISSUE TITLE
Unable to Search Mod Vault

### ISSUE THREAD

**Issue by TheMak** [FAForever/server#15]
When performing a server search the lobby just hangs with the small window titled Connecting to Statistics Server poping up with the message Connecting to Server.

In my forever.log, these are the last lines that deal with the query:
```
2015-02-20 09:08:33,342 DEBUG    faf.modvault         Updating visibilities with sort 'date' and visibility 'all'
2015-02-20 09:08:56,108 ERROR    faf.secondaryServer  Operation timed out while waiting for info.
```
The ‘ERROR’ line just constantly repeats and in a short time just make the log file over 2GB. 

Lobby version: 0.10.122

The last time I had tried this search and was certain it was successful was on Saturday February 7, 11:47 pm EST (Sunday February 8, 4:47 am GMT).  This was the last time I had downloaded a mod from the vault.

The first time that I had run into this issue was Wednesday, February 18, 11:00am EST (Wednesday, February 18, 4:00pm GMT).

Looking at the Commits history, all the changes during that time period were under Server, faf.webapp, and fa.

Edit:

The search function works when I enter some text, but it hangs when I leave the text field blank.  That is I want the query to return the latest mods uploaded to the vault.

**Comment by TheMak** [comment]
I did some more testing.  If you enter in at least two letters into the search field, select Upload Date from the pull down menu, and then hit Server Search, you will get back a query.

If you then delete the text in the search field, you will now get a list of mods with the most recently uploaded first.

Just an FYI.

**Comment by TheMak** [comment]
Whatever changes have been made has made the problem worse.

The workaround that I described above for listing the most recent uploaded mods to the vault no longer works.  It instead continues to use the initial top 100 rated mods that pop up when you open the mod vault.  In order to find a mod now you need to know at least a few characters in the right order for it to show up in the results lists.

**Comment by Sheeo** [comment]
No changes were made to the mod vault since october.

**Comment by TheMak** [comment]
That is correct, the issue is somehow related to the query search.

Before the date I had posted above, if you leave the text field blank, set the pull down menu to Upload Date, and then click on Server Search it would return a list of the most resent uploaded mods.

If you try this step today, the search will hang the lobby and the error that I had listed above will spam the log.

It is as if the server does not know how to handle a query that is blank.  If you enter in at least two letters it will return a query (will search through the mod name and the author field) and present the results with the most recent uploaded mod first.  If you go back to the text field and erase the two characters you entered in you will now see a list that contains the return result mixed with the initial list of most rated mods, sorted by date of upload with the most recent at the top of the list.

**Comment by TheMak** [comment]
I am trying to pin down the issue, but I am not following to well the programming language.

In `lobby/src/modvault/__init__.py`
at line 178
there is the function: def search(self):
This sends the contents of the search query textfield to the function modvault_search

I think there may be a problem here if there is no text (blank) entered.  It has worked in the past, just do not know why not now. 


Continuing on..

modvault_search is located in `server/replays/replayServerThread.py`
line 80, that begins with
def command_modvault_search(self, message):

In that function there were two commits made on February 4 that was labeled Convert to Python 3.  This commit affected a number of files so it is a bit hard to find this file amongst all the changes.

Anyways, in `server/replays/replayServerThread.py` the two changes were as follows (there was a recent commit on March 16 so I am using the new line numbers):

on line 129:
while next(query):

it was:
while query.next():

and on line 148:
thumbstr = config['global']['content_url'] + "vault/mods_thumbs/" + urllib.parse.quote(icon)

it was:
thumbstr = config['global']['content_url'] + "vault/mods_thumbs/" + urllib2.quote(icon)


I think the issue may be at line 129. I have been trying to find out if there is an order or logic issue, but I cannot seem to find one.

In all my testing on the search field, I have managed to crash the client when entering this list of two characters and running a server search:
al, an, as, im, ol, on, us

Also any single digit number entered in the text field works fine for the query.

I have had the query return no results on a number of two character searches such as `aq` and the client did not crash.

So the client is able to send back a query and able to return an empty query.

So I am puzzled as to why a query with no text or the list of two characters from above will hang the client.

In my testing the two characters I used was a vowel and any other letter.  I ran through every combination of that for each vowel (a, e, i, o, u, and y).

Either way, not being able to see the latest mods that have been uploaded sucks.
